src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/SourceTextDiffer.cs

```diff
@@ -0,0 +1,143 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    internal class SourceTextDiffer : TextDiffer
+    {
+        private bool _lineDiffOnly;
+
+        private SourceTextDiffer(SourceText oldText, SourceText newText, bool lineDiffOnly)
+        {
+            OldText = oldText;
+            NewText = newText;
+            _lineDiffOnly = lineDiffOnly;
+        }
+
+        private SourceText OldText { get; }
+
+        private SourceText NewText { get; }
+
+        protected override int OldTextLength => _lineDiffOnly ? OldText.Lines.Count : OldText.Length;
+
+        protected override int NewTextLength => _lineDiffOnly ? NewText.Lines.Count : NewText.Length;
+
+        public static IReadOnlyList<TextChange> GetMinimalTextChanges(SourceText oldText, SourceText newText, bool lineDiffOnly = true)
```

Comment:
If we ever want to squeeze more perf out of this we could make it not return `TextChange`s and instead have it return a type that looked like `TextChange`'s but utilized `Span`s to pull `string` content from the original source text's. That way you wouldn't re-allocate large strings for the `.NewText` property of each `TextChange` 😄. That being said it'd probably require a custom JsonConverter to make it possible.

Reply:
> re-allocate large strings for the .NewText property

Yup, I bet this largely contributes to the amount of allocation. This is a good one to keep in mind if we ever need to revisit the perf here in the future. Another opportunity here is to optimize the `ContentEquals` implementation for line content comparison so we don't have to compare full line content each time (we can maybe build a hash for each line and compare that). And I'm sure there are a lot of per improvement opportunities here.